File: src/lp_value_calc.ts
import { ethers, ZeroAddress, MaxUint256 } from 'ethers';
import { JsonRpcProvider } from 'ethers';
import { Contract } from 'ethers';
import { Decimal } from 'decimal.js';

interface PositionValue {
    tokenPrice: number;
    tokenXPrice: number;
    tokenYPrice: number;
    positionValue: number;
    unclaimedFees: number;
    totalValue: number;
}

export class V3LpValue {
    private provider: JsonRpcProvider;
    private apiKey: string;
    public priceLower: number = 0;
    public priceUpper: number = 999999999999;

    private readonly tokenAbi = [
        {
            constant: true,
            inputs: [],
            name: "decimals",
            outputs: [{ name: "", type: "uint8" }],
            payable: false,
            stateMutability: "view",
            type: "function"
        }
    ];

    constructor(rpcUrl: string, apiKey: string) {
        this.provider = new JsonRpcProvider(rpcUrl);
        this.apiKey = apiKey;
    }

    // Calculate liquidity
    private calculateV3Liquidity(
        amountX: number,
        amountY: number,
        sqrtPriceCurrent: number,
        sqrtPriceLower: number,
        sqrtPriceUpper: number
    ): number {
        if (sqrtPriceCurrent <= sqrtPriceLower) {
            // Price below range
            return amountX * (sqrtPriceUpper * sqrtPriceLower) / (sqrtPriceUpper - sqrtPriceLower);
        } else if (sqrtPriceCurrent >= sqrtPriceUpper) {
            // Price above range
            return amountY / (sqrtPriceUpper - sqrtPriceLower);
        } else {
            // Price within range
            const lx = amountX * (sqrtPriceUpper * sqrtPriceCurrent) / (sqrtPriceUpper - sqrtPriceCurrent);
            const ly = amountY / (sqrtPriceCurrent - sqrtPriceLower);
            return Math.min(lx, ly);
        }
    }

    // Calculate USD value of liquidity
    private calculatePositionValue(
        liquidity: number,
        currentPrice: number,
        tokenXPriceUsd: number,
        tokenYPriceUsd: number
    ): number {
        const amountX = liquidity / Math.sqrt(currentPrice);
        const amountY = liquidity * Math.sqrt(currentPrice);
        return (amountX * tokenXPriceUsd) + (amountY * tokenYPriceUsd);
    }

    // Get token amounts and unclaimed fees from LP position
    private async getPositionAmounts(
        positionId: number,
        nftContract: Contract,
        poolContract: Contract,
        blockTag?: number
    ): Promise<[number, number, number, number]> {
        const overrides = blockTag ? { blockTag } : {};
        const position = await nftContract.positions(positionId, overrides);
        
        const liquidity = new Decimal(position.liquidity.toString());
        const tickLower = Number(position.tickLower);
        const tickUpper = Number(position.tickUpper);
        
        const slot0 = await poolContract.slot0(overrides);
        const sqrtPriceX96 = new Decimal(slot0.sqrtPriceX96.toString());
        const currentTick = Number(slot0.tick);
        
        const sqrtPriceCurrent = sqrtPriceX96.div(new Decimal(2).pow(96));
        const sqrtPriceLower = new Decimal(1.0001).pow(tickLower / 2);
        const sqrtPriceUpper = new Decimal(1.0001).pow(tickUpper / 2);

        let amount0: Decimal, amount1: Decimal;
        
        if (currentTick < tickLower) {
            amount0 = liquidity.mul(sqrtPriceLower.pow(-1).sub(sqrtPriceUpper.pow(-1)));
            amount1 = new Decimal(0);
        } else if (currentTick > tickUpper) {
            amount0 = new Decimal(0);
            amount1 = liquidity.mul(sqrtPriceUpper.sub(sqrtPriceLower));
        } else {
            amount0 = liquidity.mul(sqrtPriceCurrent.pow(-1).sub(sqrtPriceUpper.pow(-1)));
            amount1 = liquidity.mul(sqrtPriceCurrent.sub(sqrtPriceLower));
        }

        // Get unclaimed fees
        const MAX_UINT128 = "0xffffffffffffffffffffffffffffffff"; // 128-bit max value

        const collectParams = {
            tokenId: positionId,
            recipient: ZeroAddress,
            amount0Max: MAX_UINT128,
            amount1Max: MAX_UINT128
        };

        try {
            const unclaimedFees = await nftContract.collect.staticCall(collectParams, overrides);
            return [
                Number(amount0.toString()),
                Number(amount1.toString()),
                Number(unclaimedFees.amount0.toString()),
                Number(unclaimedFees.amount1.toString())
            ];
        } catch (e) {
            console.error(`Error getting unclaimed fees at block ${blockTag || 'latest'}:`, e);
            return [
                Number(amount0.toString()),
                Number(amount1.toString()),
                0,
                0
            ];
        }
    }

    // Get token price from pool
    private async getTokenPriceFromPool(
        poolContract: Contract,
        isToken0: boolean = true,
        blockTag?: number
    ): Promise<[number, number, number]> {
        const overrides = blockTag ? { blockTag } : {};
        const token0 = await poolContract.token0(overrides);
        const token1 = await poolContract.token1(overrides);
        
        const token0Contract = new Contract(token0, this.tokenAbi, this.provider);
        const token1Contract = new Contract(token1, this.tokenAbi, this.provider);
        
        const decimals0 = Number(await token0Contract.decimals(overrides));
        const decimals1 = Number(await token1Contract.decimals(overrides));
        
        const slot0 = await poolContract.slot0(overrides);
        const sqrtPriceX96 = new Decimal(slot0.sqrtPriceX96.toString());
        
        // Avoid using pow, use multiplication instead
        const Q96 = new Decimal(2).pow(96);
        const sqrtPrice = sqrtPriceX96.div(Q96);
        const price = sqrtPrice.mul(sqrtPrice);
        
        // Use string manipulation for precision adjustment
        const decimalDiff = decimals0 - decimals1;
        const decimalAdjustment = new Decimal('1' + '0'.repeat(Math.abs(decimalDiff)));
        const actualPrice = decimalDiff >= 0 
            ? price.mul(decimalAdjustment)
            : price.div(decimalAdjustment);
        
        return [Number(actualPrice.toString()), decimals0, decimals1];
    }

    // Get contract ABI
    private async getContractAbi(contractAddress: string, chain: 'arbitrum' | 'eth'): Promise<any> {
        const endpoint = chain === 'arbitrum' 
            ? "https://api.arbiscan.io/api"
            : "https://api.etherscan.io/api";
        
        const params = new URLSearchParams({
            module: 'contract',
            action: 'getabi',
            address: contractAddress,
            apikey: this.apiKey
        });

        try {
            const response = await fetch(`${endpoint}?${params}`);
            const data = await response.json();
            if (data.status === '1') {
                return JSON.parse(data.result);
            } else {
                // throw new Error(`Error getting ABI: ${contractAddress} ${data.result}`);
                console.log(`Error getting ABI: ${contractAddress} ${data.result}`);
            }
        } catch (e) {
            console.error("Error:", e);
            return null;
        }
    }

    // Calculate LP position value
    public async calculateLpValue(
        positionManagerAddress: string,
        poolAddress: string,
        positionId: number,
        blockNumber?: number,
        poolXUAddress: string = '',
        poolYUAddress: string = ''
    ): Promise<PositionValue & { blockNumber: number }> {
        // Get contract ABI
        // let positionAbi = await this.getContractAbi(positionManagerAddress, 'arbitrum');
        // if (!positionAbi) {
        //     positionAbi = require('../asset/abi/ERC20/position_manager_address.json');
        // }
        
        // let poolAbi = await this.getContractAbi(poolAddress, 'arbitrum');
        // if (!poolAbi) {
        //     poolAbi = require('../asset/abi/ERC20/pool_address.json');
        // }

        let positionAbi = require('../asset/abi/ERC20/position_manager_address.json');
        let poolAbi = require('../asset/abi/ERC20/pool_address.json');

        const nftContract = new Contract(positionManagerAddress, positionAbi, this.provider);
        const poolContract = new Contract(poolAddress, poolAbi, this.provider);
        
        const [amountX, amountY, feeX, feeY] = await this.getPositionAmounts(
            positionId,
            nftContract,
            poolContract,
            blockNumber
        );
        
        const [tokenPrice, decimals0, decimals1] = await this.getTokenPriceFromPool(
            poolContract,
            true,
            blockNumber
        );

        const tokenXUPrice = poolXUAddress ? 
            (await this.getTokenPriceFromPool(
                new Contract(poolXUAddress, poolAbi, this.provider),
                true,
                blockNumber
            ))[0] : 
            tokenPrice;
            
        const tokenYUPrice = poolYUAddress ? 
            (await this.getTokenPriceFromPool(
                new Contract(poolYUAddress, poolAbi, this.provider),
                true,
                blockNumber
            ))[0] : 
            1;

        const sqrtPriceCurrent = Math.sqrt(tokenPrice);
        const sqrtPriceLower = Math.sqrt(this.priceLower);
        const sqrtPriceUpper = Math.sqrt(this.priceUpper);
        
        const liquidity = this.calculateV3Liquidity(
            amountX / (10 ** decimals0),
            amountY / (10 ** decimals1),
            sqrtPriceCurrent,
            sqrtPriceLower,
            sqrtPriceUpper
        );
        
        const positionValue = this.calculatePositionValue(
            liquidity,
            tokenPrice,
            tokenXUPrice,
            tokenYUPrice
        );
        
        const feeValue = (feeX * tokenXUPrice / (10 ** decimals0)) + 
                        (feeY * tokenYUPrice / (10 ** decimals1));
        
        const currentBlock = blockNumber || await this.provider.getBlockNumber();
        
        return {
            blockNumber: currentBlock,
            tokenPrice,
            tokenXPrice: tokenXUPrice,
            tokenYPrice: tokenYUPrice,
            positionValue,
            unclaimedFees: feeValue,
            totalValue: positionValue + feeValue
        };
    }
}

// Usage example
async function main() {
    // Load configuration
    const config = require('./config.json');
    
    // Set proxy if configured
    if (config.proxy.http && config.proxy.https) {
        process.env.http_proxy = config.proxy.http;
        process.env.https_proxy = config.proxy.https;
    }
    
    // Create client instance
    const client = new V3LpValue(config.network.rpc, config.network.api_key);
    client.priceLower = config.position.price_lower;
    client.priceUpper = config.position.price_upper;
    
    // Calculate LP value
    const result = await client.calculateLpValue(
        config.contracts.position_manager,
        config.contracts.pool,
        config.position.id,
        config.query.block_number
    );
    
    // Print results
    console.log(`Block Number: ${result.blockNumber}`);
    console.log(`Pool Price: ${result.tokenPrice.toFixed(4)}`);
    console.log(`Token X Price: ${result.tokenXPrice.toFixed(4)}`);
    console.log(`Token Y Price: ${result.tokenYPrice.toFixed(4)}`);
    console.log(`Position Value: $${result.positionValue.toFixed(2)}`);
    console.log(`Unclaimed Fees: $${result.unclaimedFees.toFixed(6)}`);
    console.log(`Total Value: $${result.totalValue.toFixed(2)}`);
}

if (require.main === module) {
    main().catch(console.error);
} 

File: src/lp_value_calc.py
from math import sqrt
from web3 import Web3
from decimal import Decimal
from numpy import inf
import json
import time
import os
import requests


class v3_lp_value:
    def __init__(self, rpc_url, api_key):
        """
        rpc_url: RPC node URL
        api_key: Arbiscan API key
        """
        self.w3 = Web3(Web3.HTTPProvider(rpc_url))
        self.api_key = api_key
        
        # Base token contract ABI
        self.token_abi = [
            {
                "constant": True,
                "inputs": [],
                "name": "decimals",
                "outputs": [{"name": "", "type": "uint8"}],
                "payable": False,
                "stateMutability": "view",
                "type": "function"
            }
        ]


    # Calculate liquidity
    def calculate_v3_liquidity(self, amount_x, amount_y, sqrt_price_current, sqrt_price_lower, sqrt_price_upper):
        """
        amount_x: Amount of token X
        amount_y: Amount of token Y
        sqrt_price_current: Square root of current price
        sqrt_price_lower: Square root of lower price bound
        sqrt_price_upper: Square root of upper price bound
        """
        if sqrt_price_current <= sqrt_price_lower:
            # Price below range
            liquidity = amount_x * (sqrt_price_upper * sqrt_price_lower) / (sqrt_price_upper - sqrt_price_lower)
        elif sqrt_price_current >= sqrt_price_upper:
            # Price above range
            liquidity = amount_y / (sqrt_price_upper - sqrt_price_lower)
        else:
            # Price within range
            lx = amount_x * (sqrt_price_upper * sqrt_price_current) / (sqrt_price_upper - sqrt_price_current)
            ly = amount_y / (sqrt_price_current - sqrt_price_lower)
            liquidity = min(lx, ly)
        
        return liquidity


    # Calculate USD value of liquidity
    def calculate_position_value(self, liquidity, current_price, token_x_price_usd, token_y_price_usd):
        """
        liquidity: Amount of liquidity
        current_price: Current pool token price
        token_x_price_usd: Token X price in USD
        token_y_price_usd: Token Y price in USD
        """
        amount_x = liquidity / sqrt(current_price)
        amount_y = liquidity * sqrt(current_price)
        
        total_value = (amount_x * token_x_price_usd) + (amount_y * token_y_price_usd)
        return total_value


    # Get token amounts and unclaimed fees from LP position
    def get_position_amounts(self, position_id, nft_contract, pool_contract, block_identifier='latest'):
        """
        position_id: NFT token ID
        nft_contract: Uniswap V3 NFT manager contract
        pool_contract: Uniswap V3 pool contract
        block_identifier: Target block number (optional)
        returns: (amount0, amount1, fee0, fee1) Token amounts and unclaimed fees
        """
        position = nft_contract.functions.positions(position_id).call(block_identifier=block_identifier)
        
        liquidity = position[7]  # Liquidity
        tick_lower = position[5]  # Lower price bound
        tick_upper = position[6]  # Upper price bound
        
        # Get current tick and price
        slot0 = pool_contract.functions.slot0().call(block_identifier=block_identifier)
        sqrt_price_x96 = slot0[0]
        current_tick = slot0[1]
        
        # Calculate price range
        sqrt_price_current = Decimal(sqrt_price_x96) / Decimal(2**96)
        sqrt_price_lower = Decimal(1.0001 ** (tick_lower/2))
        sqrt_price_upper = Decimal(1.0001 ** (tick_upper/2))
        
        # Calculate token amounts
        if current_tick < tick_lower:
            amount0 = liquidity * (1/sqrt_price_lower - 1/sqrt_price_upper)
            amount1 = 0
        elif current_tick > tick_upper:
            amount0 = 0
            amount1 = liquidity * (sqrt_price_upper - sqrt_price_lower)
        else:
            amount0 = liquidity * (1/sqrt_price_current - 1/sqrt_price_upper)
            amount1 = liquidity * (sqrt_price_current - sqrt_price_lower)
        
        # Get unclaimed fees
        MAX_UINT128 = 2**128 - 1
        
        # Construct collect parameters
        collect_params = {
            'tokenId': position_id,
            'recipient': '0x0000000000000000000000000000000000000000',  # Address doesn't matter as this is a static call
            'amount0Max': MAX_UINT128,
            'amount1Max': MAX_UINT128
        }
        
        # Use callStatic to simulate collect call with block parameter
        try:
            unclaimed_fees = nft_contract.functions.collect(collect_params).call(block_identifier=block_identifier)
            fee0, fee1 = unclaimed_fees[0], unclaimed_fees[1]
        except Exception as e:
            print(f"Error getting unclaimed fees at block {block_identifier}: {e}")
            fee0, fee1 = 0, 0
        
        return float(amount0), float(amount1), float(fee0), float(fee1)


    # Get token price from pool
    def get_token_price_from_pool(self, pool_contract, is_token0=True, block_identifier='latest'):
        """
        pool_contract: Uniswap V3 pool contract
        is_token0: Whether to get token0 price (True for token0 price, False for token1)
        block_identifier: Target block number (optional)
        """
        token0 = pool_contract.functions.token0().call(block_identifier=block_identifier)
        token1 = pool_contract.functions.token1().call(block_identifier=block_identifier)
        
        token0_contract = self.w3.eth.contract(address=token0, abi=self.token_abi)
        token1_contract = self.w3.eth.contract(address=token1, abi=self.token_abi)
        
        decimals0 = token0_contract.functions.decimals().call(block_identifier=block_identifier)
        decimals1 = token1_contract.functions.decimals().call(block_identifier=block_identifier)
        
        slot0 = pool_contract.functions.slot0().call(block_identifier=block_identifier)
        sqrt_price_x96 = slot0[0]
        
        price = (Decimal(sqrt_price_x96) / Decimal(2**96)) ** 2
        decimal_adjustment = Decimal(10 ** (decimals0 - decimals1))
        actual_price = price * decimal_adjustment
        
        return float(actual_price), decimals0, decimals1


    # Get contract ABI
    def get_contract_abi(self, contract_address, chain):
        """
        contract_address: Contract address
        """
        if chain == 'arbitrum':
            end_point = "https://api.arbiscan.io/api"
        elif chain == 'eth':
            end_point = "https://api.etherscan.io/api"
        
        params = {
            'module': 'contract',
            'action': 'getabi',
            'address': contract_address,
            'apikey': self.api_key
        }
        
        try:
            response = requests.get(end_point, params=params)
            response_json = response.json()
            if response_json['status'] == '1':
                return json.loads(response_json['result'])
            else:
                raise Exception(f"Error getting ABI: {contract_address} {response_json['result']}")
        except Exception as e:
            print(f"Error: {e}")
            return None


    # Calculate LP position value
    def main(self, position_manager_address, pool_address, position_id, 
             block_number=None, pool_x_u_address='', pool_y_u_address=''):
        """
        position_manager_address: Uniswap V3 Position Manager address
        pool_address: Pool contract address
        position_id: NFT Position ID
        block_number: Target block number (optional)
        pool_x_u_address: Token0-USDC pool address
        pool_y_u_address: Token1-USDC pool address
        """
        position_manager_address = self.w3.to_checksum_address(position_manager_address)
        pool_address = self.w3.to_checksum_address(pool_address)

        # Create block parameter
        block_identifier = block_number if block_number is not None else 'latest'

        # Get contract ABI
        with open(r'asset\abi\ERC20\position_manager_address.json', 'r') as f:
            position_abi = json.load(f)
        with open(r'asset\abi\ERC20\pool_address.json', 'r') as f:
            pool_abi = json.load(f)

        nft_contract = self.w3.eth.contract(address=position_manager_address, abi=position_abi)
        pool_contract = self.w3.eth.contract(address=pool_address, abi=pool_abi)
        
        # Get token amounts and unclaimed fees
        amount_x, amount_y, fee_x, fee_y = self.get_position_amounts(
            position_id, 
            nft_contract, 
            pool_contract,
            block_identifier
        )
        
        token_price, decimals0, decimals1 = self.get_token_price_from_pool(
            pool_contract,
            block_identifier=block_identifier
        )

        if pool_x_u_address:
            pool_x_contract = self.w3.eth.contract(address=pool_x_u_address, abi=pool_abi)
            token_x_u_price = self.get_token_price_from_pool(pool_x_contract, block_identifier=block_identifier)[0]
        else:
            token_x_u_price = token_price
            
        if pool_y_u_address:
            pool_y_contract = self.w3.eth.contract(address=pool_y_u_address, abi=pool_abi)
            token_y_u_price = self.get_token_price_from_pool(pool_y_contract, block_identifier=block_identifier)[0]
        else:
            token_y_u_price = 1

        sqrt_price_current = sqrt(token_price)
        sqrt_price_lower = sqrt(self.price_lower)
        sqrt_price_upper = sqrt(self.price_upper)
        
        # Calculate main liquidity value
        liquidity = self.calculate_v3_liquidity(
            amount_x / (10 ** decimals0),
            amount_y / (10 ** decimals1), 
            sqrt_price_current,
            sqrt_price_lower,
            sqrt_price_upper
        )
        
        position_value = self.calculate_position_value(
            liquidity,
            token_price,
            token_x_u_price,
            token_y_u_price
        )
        
        # Calculate value of unclaimed fees
        fee_value = (fee_x * token_x_u_price / (10 ** decimals0)) + (fee_y * token_y_u_price / (10 ** decimals1))
        total_value = position_value + fee_value
        
        return {
            'block_number': block_number if block_number is not None else self.w3.eth.block_number,
            'token_price': token_price,
            'token_x_price': token_x_u_price,
            'token_y_price': token_y_u_price,
            'position_value': position_value,
            'unclaimed_fees': fee_value,
            'total_value': total_value
        }


if __name__ == "__main__":
    # Load configuration
    with open('config.json', 'r') as f:
        config = json.load(f)
    
    # Set proxy if configured
    if config['proxy']['http'] and config['proxy']['https']:
        os.environ["http_proxy"] = config['proxy']['http']
        os.environ["https_proxy"] = config['proxy']['https']
    
    # Create client instance
    client = v3_lp_value(config['network']['rpc'], config['network']['api_key'])
    client.price_lower = config['position']['price_lower']
    client.price_upper = config['position']['price_upper']
    
    # Calculate LP value
    result = client.main(
        position_manager_address=config['contracts']['position_manager'],
        pool_address=config['contracts']['pool'],
        position_id=config['position']['id'],
        block_number=config['query']['block_number']
    )
    
    # Print results
    print(f"Block Number: {result['block_number']}")
    print(f"Pool Price: {result['token_price']:.4f}")
    print(f"Token X Price: {result['token_x_price']:.4f}")
    print(f"Token Y Price: {result['token_y_price']:.4f}")
    print(f"Position Value: ${result['position_value']:.2f}")
    print(f"Unclaimed Fees: ${result['unclaimed_fees']:.6f}")
    print(f"Total Value: ${result['total_value']:.2f}")
